### PR TITLE
SRE-99 Build: Move Bullseye to EL-8

### PR DESCRIPTION
**Pull request by JohnMalmberg**
Change the Bullseye stages to be EL-8

Skip-build-ubuntu20-rpm: true
Skip-build-centos7-gcc: true
Skip-bullseye: false
Skip-fnbullseye: false
Skip-func-hw-test: true
Skip-nlt: true
Skip-unit-test-memcheck: true

Signed-off-by: John E Malmberg <john.e.malmberg@intel.com>

Required-githooks: true

### CODE REVIEW

File: utils/docker/Dockerfile.el.8
Comment (ashleypittman) [inline utils/docker/Dockerfile.el.8:96]: This should be in utils/scripts/install-el8.sh obviously.

Comment (brianjmurrell) [reply utils/docker/Dockerfile.el.8:96]: Should it?  It's only needed when doing the Bullseye build.

**Review by brianjmurrell — COMMENTED** [review]
Why is it that we don't have this enabled by default?  Is that a/the goal of getting it working on EL8 here?  Should that be included in this PR?

Why is all of the HTML that Bullseye produces wrapped up in a `.zip` rather than just being archived natively?  If we did the latter, the HTML reports would be visible directly in Jenkins rather than having to download ZIPs, unpack them, etc.

File: utils/docker/Dockerfile.el.8
Comment (ashleypittman) [reply utils/docker/Dockerfile.el.8:96]: Yes, we have lots of packages there which are only needed for Java builds etc.

Comment (brianjmurrell) [reply utils/docker/Dockerfile.el.8:96]: But the Java build is not conditional and defaulted to off, is it?

I don't really feel all that strongly about this, but just trying to avoid unnecessary bloat (for (al)most (all)) everyone.

But really, if `zip` is being included to create the `.zip`'ed artifact, maybe it can be gotten rid of by addressing my other question about the usefulness of zipping up the HTML.

Comment (ashleypittman) [reply utils/docker/Dockerfile.el.8:96]: Having all the packages in one place makes things simpler but yes it does mean potentially you're installed things that you don't need to, however the alternative is many install scripts with associated maintenance burden, plus the extra run-time and failure points when running the dockerfile.

That said the Java build does it's own downloads of anything it needs from the java side anyway.

+1 on the zip comments, we shouldn't really be zipping things up and if we do need to bundle them then tar + gz would be preferable if bullseye has that option.

Comment (brianjmurrell) [reply utils/docker/Dockerfile.el.8:96]: +1 on your +1 including the `.tar.gz` alternative to `.zip`.  But we really need to examine the usefulness of any kind of bundling of the HTML into single file.

Comment (JohnMalmberg) [reply utils/docker/Dockerfile.el.8:96]: A few years ago, Jenkins put security fixes for what its web server is allowed to serve.  The HTML content generated for by Bullseye for the Clover plugin is blocked by those fixes and can not be displayed.

Zip archive format was used because the .tar.gz format is not easily used on Windows platforms, and also supported on the platforms.

The zipping of the web site for local download and display is  workaround until SRE can work on a better solution.

Comment (JohnMalmberg) [reply utils/docker/Dockerfile.el.8:96]: Bullseye is a commercial product and requires a commercial license key to install/use.
The dockerfile section that uses BULLSEYE can not be used outside of our environment, as in third-parties using the Dockerfile, or by github actions.

CI had bullseye builds off by default because CI runs in the past had high failure rates on Unit testing with Bullseye.  Most seemed to be from the instrumentation slowing down some tests or otherwise affecting the run.   The Bullesye instrumented runs also added considerable more time to CI runs than they appear to be doing now.

CI Functional testing requires that the RPMs be built with Bullseye instrumentation built in, and those instrumented RPMs binaries do not appear to be in their own package with their own names, so enabling that changes both the RPM artifacts and all functional testing to be run with the instrumentation on.

**Review by JohnMalmberg — COMMENTED** [review]
Need a decision on moving the code from the Dockerfile.

Still do not know why the NLT testing is failing, as nothing in this PR should be affecting the NLT tests.

**Comment by ashleypittman** [comment]
The NLT failures are all valgrind related so one of the suppressions isn't matching for some reason.  I can't see why anything in this PR would have an effect but if we can work that out then adding an appropriate suppression should be easy.

**Comment by JohnMalmberg** [comment]
I need to merge this with current master and get a run with default of bullseye being off

**Comment by JohnMalmberg** [comment]
Tripped a pipeline-lib bug with the bullseye tests, so re-running with fixed pipeline-lib.

On the NLT issue, quick-build is on.  Have not traced down why that is no longer enabled for the bullseye builds.
Since I was assuming that bullseye builds were going to continue to be only on special PRs, I was not planning on fixing that for this ticket.

File: utils/docker/Dockerfile.el.8
Comment (ashleypittman) [reply utils/docker/Dockerfile.el.8:96]: If you're going to add the dnf install step here then you need a "dnf clean all" command after it to remove any temporary files.

File: Jenkinsfile
Comment (brianjmurrell) [inline Jenkinsfile:19]: This will obviously have to be removed.

File: utils/docker/Dockerfile.el.8
Comment (brianjmurrell) [reply utils/docker/Dockerfile.el.8:96]: > A few years ago, Jenkins put security fixes for what its web server is allowed to serve. The HTML content generated for by Bullseye for the Clover plugin is blocked by those fixes and can not be displayed.

Have we re-evaluated this recently to see if anything has changed for the positive in Jenkins?  It should be pretty trivial to archive what is usually put into the _zip_ to see what the rendering looks like currently.

**Review by brianjmurrell — APPROVED** [review]

**Review by dinghwah — DISMISSED** [review]

**Comment by JohnMalmberg** [comment]
The pipeline-lib PR-330 needs to be landed before I can fixup the Jenkinsfile here.

**Review by dinghwah — DISMISSED** [review]
el8 bullseye build passed, hw-small test.cov file generated.

File: Jenkinsfile
Comment (brianjmurrell) [inline Jenkinsfile:172]: Why do we want to remove this?

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:4]: Interesting that we have different VS Code experiences here.

File: Jenkinsfile
Comment (brianjmurrell) [reply Jenkinsfile:19]: This still needs to be removed.

**Review by brianjmurrell — CHANGES_REQUESTED** [review]

File: Jenkinsfile
Comment (JohnMalmberg) [reply Jenkinsfile:172]: Somehow this PR ended up with two entries for this parameter, so I removed one of them.

**Review by brianjmurrell — APPROVED** [review]

**Review by dinghwah — APPROVED** [review]

**Comment by dinghwah** [comment]
@JohnMalmberg 
From the latest log, all test passed, However, there is no bullseye build, Could you verify?

**Comment by JohnMalmberg** [comment]
The last run was with no commit pragmas, the default condition,.
Default is no bullseye instrumented runs are done.
While it may be possible to have the bullseye enabled for the Unit tests, we must not have bullseye enabled for functional tests for master builds, and having bullseye enabled for PRs on functional tests may affect the results of those tests.

**Comment by dinghwah** [comment]
@JohnMalmberg  Bullseye was working on EL-7,  this PR "move Bullseye to EL-8" does not have Bullseye enabled?

**Review by dinghwah — CHANGES_REQUESTED** [review]
Please run tests with Bullseye enable

File: ci/rpm/build.sh
Comment (phender) [inline ci/rpm/build.sh:27]: Should this default be `el-7`?
```suggestion
: "${REPO_SPEC:='el-7'}"
```

**Review by phender — COMMENTED** [review]
Assuming https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-9482/71/pipeline is the build with bullseye enabled (and therefore verifying these changes work), why does the 'Bullseye Report on EL 8' stage appear green in BO with steps marked with red X's?  Is this some sort of override where the coverity file is expected to be missing for the stages not run?   Is there a way to handle this w/o the red X's?

File: ci/rpm/build.sh
Comment (JohnMalmberg) [reply ci/rpm/build.sh]: That is only used for debugging outside of Jenkins, but yes, it should be el-7.
Fixed in the current push.

**Review by JohnMalmberg — COMMENTED** [review]
The bullseye report has the red sections because I did not run two of the hardware stages.
Getting the functional hardware stages to pass will be the job of the Bullseye test maintainer as that is new functionality added with this PR.

Bullseye data needs to be passed to the report stage via stashes, and the report stage does not have good way of knowing which stages were run with Bullseye instrumentation, so it needs to do a test unstash and catch the exception if if fails.   The caught exception always shows up as a red step in the Jenkins stage log.

**Review by phender — APPROVED** [review]

**Review by dinghwah — APPROVED** [review]

**Comment by JohnMalmberg** [comment]
Merged with current master to try to get a test run with no failures.

**Comment by JohnMalmberg** [comment]
Finally got a passing run.

**Pull request merged**
